### DISCUSSION TITLE
Trying to colour items on a custom map

### DISCUSSION THREAD

**Discussion by rachellawson** [osmandapp/OsmAnd#19102] [Maps Related]
I have a custom map, created from a fairly large geojson dataset.

I’ve imported the data into JOSM and the data exists as ways, an example of which looks like this:

<img width="546" alt="Screenshot 2024-02-13 at 19 17 19" src="https://github.com/osmandapp/OsmAnd/assets/807248/bbb0fab7-b240-481f-aefe-79ffe96bbf82">

I then used OsmAndMapCreator to turn that into an .obf and loaded into Osmand.

I have then created my own custom version of the “off-road” map renderer that includes the following added snippet:

```
<!-- Rachel trying to highlight proposed routes. —>
<case tag="highway" value="proposed"> 
  <case minzoom="6" strokeWidth_4="14" strokeWidth_5="16"/> 
  <apply color_4="#aaaaaa" pathEffect_5="10_3"/> 
  <case additional="class=full-access" color_5="#00cc00" color_4="#888888" pathEffect_5="10_3"/>
  <case additional="color=red">
    <apply color_5="#ff0000" color_4="#888888" pathEffect_5="10_3"/>
  </case>
  <case tag="color" value="blue">
    <apply color_5="#0000ff" color_4="#888888" pathEffect_5="10_3”/>
  </case>
  <case tag="color" value="grey">
    <apply color_5="#dddddd" color_4="#888888" pathEffect_5="10_3"/>
  </case>
</case>
```

and I was hoping for a grey/color dotted line - either green/red/blue/grey

but I get this:

![IMG_0326](https://github.com/osmandapp/OsmAnd/assets/807248/4c938cf7-e5e2-4a02-ab5e-9a867960127c)

Pretty but the grey lines are not highlighted in the specific green/red/blue/grey colours I’d hoped for.

I’m pretty sure I’m getting something wrong in the rendering but not sure what?

**Comment by osmandlover** [comment]
Hi Rachel

Usually when this happens it is because you are relying on a tag that is present in the *.osm data but is not present in the *.obf file.

OsmAndMapCreator uses an XML file (the default one can be found at `/resources/obf_creation/rendering_types.xml`) to make a lot of transformations on the *.osm file. It chucks out a lot of data that isn't used by the renderer. I think it is done to make the maps smaller and faster to render, however it's a bit of a PITA and a bit of a pity that osmand can't use raw osm data which would be a lot simpler!

In your case you are using a custom tag "color" to specify some additional style. This tag is not passed through into the obf by default, looking at the default render_types.xml.

You have two options:

1. Use a different tag/property that IS passed into the obf. This is usually the easier AND better way because then maps made by third parties and downloaded automatically by OsmAnd will still work, and keeps the rendering style separate from the data itself, which is cleaner. For example, if "red" track specifies a surface type, then better to rather look at that. If it's a very specific use case and they are routes or names, you might be able to use the name field to decide which color to render. There are also tags like "color_red" specified in the `render_types.xml` file, so that could also work?

2. If there isn't an existing tag you can use to distinguish the ways, you will need to add the color tag to the rendering_types.xml as a tag that needs to be passed into the obf. This is a little more tricky as you will need to work with the arcane syntax of the rendering_types.xml custom language. And you will need to add the path to the rendering types file in OsmAndMapCreator > Window > Settings > Path to rendering_types.xml. Then rebuild any maps that you wish to have your custom styling (and keep them up to date manually - so you will probably want to do that programmatically - which is also possible). 

Let me know if you need more help with either of these two methods, I do both of these regularly!

**Comment by rachellawson** [comment]
Aha! This was the level of detail I needed to figure this out - thank you!

By adding a few directives to a copy of rendering_types.xml:

```
<type tag="class" value="full-access" minzoom="9" additional="true"/>
<type tag="class" value="partial-access" minzoom="9" additional="true"/>
<type tag="class" value="temporary_tro" minzoom="9" additional="true"/>
<type tag="class" value="restricted" minzoom="9" additional="true"/>
```

and altering my rendering a little to more like:

```
<apply_if additional="class=full-access" color_5="#00ff00" color_4="#888888" pathEffect_5="10_3"/>
<apply_if additional="class=partial-access" color_5="#00cc00" color_4="#888888" pathEffect_5="10_3"/>
<apply_if additional="class=temporary_tro" color_5="#0000cc" color_4="#888888" pathEffect_5="10_3"/>
<apply_if additional="class=restricted" color_5="#cc0000" color_4="#888888" pathEffect_5="10_3"/>
```

I have exactly the map I needed!!!

![IMG_0328](https://github.com/osmandapp/OsmAnd/assets/807248/2349df2a-2978-4141-a74c-658843923729)

Again, thank you so much!!

**Reply by EugeneZmeuk** [reply]
@rachellawson Hi Rachel
I can propose you make the Custom plugin for any user (it will be simply to download and use your map data on OsmAnd)
For the Android version (nightly) there is the Custom plugin list (iOS users of OsmAnd can download custom plugins by clicking the link.):
![image](https://github.com/osmandapp/OsmAnd/assets/44466116/63f3df60-4f1d-4772-95da-a888e832a25b)

**Reply by rachellawson** [reply]
As it stands today, the TRF don't give permission for the data to be used in this way so I can't really distribute it. I'm working on them 😉

Honestly, helping more motorised trail users stick to the legal trails is good for us all. I'm sure I'll be successful but the timing of my success is anyone's guess.

**Comment by yet-another-steve** [comment]
On a side note that looks like an UK OS style.  s that home cooked or from somewhere?
Thx.

**Reply by rachellawson** [reply]
To be honest, the one above is just the “Offroad” theme. I do have a homegrown OS-style, one, though.

<table style="width:100%">
  <tr>
<th><img src="https://github.com/osmandapp/OsmAnd/assets/807248/fd38922c-c1cd-4937-b4ef-26cd776bf5e9"  width="300" height="auto"/> </td>
  </tr>
</table>

**Reply by yet-another-steve** [reply]
Nice.  If you could add public_footpath, BOAT, and bridleway colouring I suspect quite a few people would be interested in an Osmand obf of the style.
